### ISSUE TITLE
different alias resolving when using requires = / requirements() or configure()

### ISSUE THREAD

**Issue by claasd** [conan-io/conan#2583]
Hi,

I encountered an error with alias resolution in conan 1.0.x, conan 1.1.0 and conan 1.1.1.
I have the following setup:
```   
   libD
     |
   libC
  /     \
libB libA
  \     /
  testing
```
for each libX, i create an alias 0.1.latest -> 0.1.0.

Now the problematic part (always using different methods how libC requires libD):

`requires = "libD/0.1.latest@demo/testing"`
works fine

```
def requirements(self):
        self.requires("libD/0.1.latest@demo/testing") 
```
gives the error   
```
Incompatible requirements obtained in different evaluations of 'requirements'
    Previous requirements: [libD/0.1.latest@demo/testing]
    New requirements: [libD/0.1.0@demo/testing]
```

```
def configure(self):
        self.requires("libD/0.1.latest@demo/testing") 
```
gives the error
```
ERROR: libC/0.1.0@demo/testing: Error in configure() method, line 7
	self.requires("libD/0.1.latest@demo/testing")
	ConanException: Duplicated requirement libD/0.1.0@demo/testing != libD/0.1.latest@demo/testing
```

I think all three methods should work.
I have attached a test-case that outlines the problem (its a python test case that should work inside conan, but py endings are forbidden).
[alias_test.py.txt](https://github.com/conan-io/conan/files/1793607/alias_test.py.txt)

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by memsharded** [comment]
Related to https://github.com/conan-io/conan/issues/2576?

**Comment by memsharded** [comment]
Hi @claasd 

As discussed https://github.com/conan-io/conan/pull/2586, regarding the current proposed fix in https://github.com/conan-io/conan/pull/2597, it is true that the first 2 cases should work, and the fix in #2597 should be enough.

Regarding the requires in ``configure()`` code, I am afraid it is not possible, and the conditional and dynamic requirements have to be defined in the ``requirements()`` method as stated all along the docs, otherwise, undefined behavior could happen. This is not something new, it had already been like this a long time. As explained in https://github.com/conan-io/conan/pull/2586, the ``requirements()`` method has some special processing to allow diamond structures to re-evaluate conditional requirements. But that can't be applied to ``configure()`` too without destroying the data.

It would help a lot if you could test the branch in #2597 against your project. It will go to 1.2 anyway, but having your feedback earlier would be great. Many thanks.

**Comment by claasd** [comment]
Hi @memsharded, Sorry that I did not give feedback, but I am working together with @wahlm, who open the PR for this issue. He tested it and for us its fine, we can move the requires from the configure method to the requires method.

Best,
Claas

**Comment by memsharded** [comment]
Hi @claasd 

No problem, thanks for telling. So then, I think we are good moving forward the PR to be released in 1.2.

Cheers.
